edx/edx-enterprise
# Add enterprise_customer field to the SystemWideEnterpriseUserRoleAssignment model

…and updates Django Admin definitions around same.

* On the `SystemWideEnterpriseUserRoleAssignment` model, adds an `enterprise_customer` FK (nullable).
* Updates the admin to show the "effective" customer in the detail view, and the explicit value in the list view.  The effective value is the deprecated way we currently determine role assignment - by implicitly assigning the role on every customer to which the user is linked.
* In the detail view/form, the "Enterprise customer" dropdown contains only customers to which the user is currently linked.
* If adding a new assignment via the admin form, the customer dropdown will contain no customers to choose from at all - the administrator must first provide a user and role, click "Save and continue editing", and then a list of linked customers will be present. The help text of the field mentions this.  Rationale: we don't want to provide a drop-down list of _every_ customer that exists in production when creating a new assignment.  The alternative would be to make this form more "dynamic", which would require more development work than provided value.

To test this locally:
* Run `./manage.py lms migrate enterprise` to apply the migration.
* Visit http://localhost:18000/admin/enterprise/systemwideenterpriseuserroleassignment/
* Change a user role assignment by having it apply to a specific enterprise to which the user is linked.
* The `effective_enterprise_customer` value should change to include only that customer.

<img width="1497" alt="role-assignment-admin-list" src="https://user-images.githubusercontent.com/2307986/104609756-60204780-5651-11eb-8377-d5a964d02f48.png">

<img width="1488" alt="role-assignment-admin-detail" src="https://user-images.githubusercontent.com/2307986/104609767-6282a180-5651-11eb-9576-addbb31aa7dc.png">

**Merge checklist:**
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.

**Post merge:**
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.
